plutus-core/exe/Main.hs

```diff
@@ -757,10 +763,14 @@ printBudgetStateTally (ExTally costs) = do
         builtinsAndCosts = Data.List.foldl f [] (H.toList costs)
 
 printBudgetState :: (Eq fun, Hashable fun, Show fun) => Cek.CekExBudgetState fun -> IO ()
-printBudgetState bs = do
-    printBudgetStateBudget $ _exBudgetStateBudget bs
+printBudgetState (CountingSt budget) =
+    printBudgetStateBudget budget
+printBudgetState (TallyingSt tally budget) = do
+    printBudgetStateBudget budget
     putStrLn ""
-    printBudgetStateTally  $ _exBudgetStateTally bs
+    printBudgetStateTally tally
+printBudgetState (RestrictingSt _) =
+    fail "This execution path was supposed to be unreachable"
```

Comment:
> This does get reached if you use `-r`

Oops, thanks for spotting that. Somehow I thought that the `Restricting` mode was `Silent`.

> [Later: I see that those numbers are in the error that occurs if the budget is exceeded, so maybe we can intercept that.]

Surely we should be able to? I'll look into that.

> FYI, I also tried running it with a really big initial budget and that causes a noticeable slowdown: testing a 60 digit prime took about 5.5s with initial limits of 10^10 but 6.5s with limits of ~10^150, and crowdfunding validation went up from 4.65ms to 5.66ms

Well, that's not surprising. Is there anything we can do about that?

Reply:
> Well, that's not surprising. Is there anything we can do about that?

Not really (although `safeint` or whatever it was called might help), but we're never going to need a bound like 10^150.  If anything it's maybe a little surprising that the slowdown's not even worse.

Reply:
We could add a max element to `ExBudget`, i.e. `data ExBudget = Budget Int Int | Unlimited` or something, and give that? Then the comparison should be fast.

Reply:
That would optimize the case efficiency of which we don't care about (unrestricted evaluation off the chain) and slowdown the primary case (restricted evaluation on the chain).

Reply:
> Printing out the remaining budget would make it a bit easier to see what's going on (although my attempt at doing that just told me that I'd used up the limits that I'd given it, where I'd naively thought that it'd say I had -5 ExCpu left or something).

Done, we now print the occurred error. Which means that in the restricting mode in case of out-of-ex failure we print the final budget twice (once via `printBudgetStateBudget` and once in the error message), not sure what to do about that, should I bother?

There was also a bug related to the final state not being recorded in case of out-of-ex failure, I fixed it.